### PR TITLE
HD wallet 1.3

### PR DESCRIPTION
**Pull request by zugzwang**
There are several issues with https://github.com/fortanix/sdkms-plugin-library/pull/6. For instance the plugin has syntax errors (cannot compile), linting errors such as trailing whitespaces, inconsistent spacing, inconsistent function and variable naming and cases, inconsistent use of string/bytes, and misleading documentation.

This version is a refactor of the plugin with the following improvements:
- More checks on signatures, for compatibility
- Bytes are used consistently
- Functions have less responsibilities, more checks, and consistent names
- Improved code documentation
- Execution errors are caught and more clear to the user
- Implementation follows https://en.bitcoin.it/wiki/BIP_0032 closely
- Tested against test vectors of BIP0032

### CODE REVIEW

File: hd-wallet/plugin.lua [hunk 1]
```diff
@@ -1,440 +1,418 @@
...
+--
 -- ## Use cases
--- 
--- The plugin can be used to
--- 
--- - Derive child key for UTXO
--- - Derive child key for ethereum
--- - Sign transaction for UTXO
--- - Sign transaction for ethereum
--- - Import wrapped private key for storage in Smartkey
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:17]: This was removed, as it was not properly implemented.

File: hd-wallet/plugin.lua [hunk 2]
```diff
@@ -1,440 +1,418 @@
...
+-- a transaction hash. The child key is transient; it only exists during the
+-- plugin execution. This version of the plugin requires the master key to be
+-- exportable. In upcoming version 2.0, this condition is removed for better
+-- security.
+--
 -- ## Use cases
--- 
--- The plugin can be used to
--- 
--- - Derive child key for UTXO
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:13]: On the new version, the child private key is not returned to the caller, so this is removed.

File: hd-wallet/plugin.lua [hunk 3]
```diff
@@ -1,440 +1,418 @@
...
-
-------------- BIP32 key structure -------------
-local key = {
-    ["version"]="",
-    ["depth"]="",       -- 1 byte depth: 0x00 for master nodes, 0x01 for level-1 derived keys, ....
-    ["index"]="",       -- 4 byte child number. This is ser32(i) for i in xi = xpar/i, with xi the key being serialized. (0x00000000 if master key)
-    ["fingerprint"]="", -- 4 byte the fingerprint of the parent's key (0x00000000 if master key)
-    ["chain_code"]="",  -- 32 byte the chain code
-    ["key"]="",         -- 33 byte key data : the public key or private key data (serP(K) for public keys, 0x00 || ser256(k) for private keys)
-    ["checksum"]=""     -- checksum of all above
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:108]: This checksum was not used or checked.

File: hd-wallet/plugin.lua [hunk 4]
```diff
@@ -1,440 +1,418 @@
...
-    return Sobject { kid = keyId }:export().value
-end
+local function deserialize_bip32(blob)
+  local key = {
+    ["version"]     = blob:slice(1, 4),   --  4 bytes
+    ["depth"]       = blob:slice(5, 5),   --  1 byte
+    ["parent_fgpt"] = blob:slice(6, 9),   --  4 bytes
+    ["index"]       = blob:slice(10, 13), --  4 bytes
+    ["chain_code"]  = blob:slice(14, 45), -- 32 bytes
+    -- ignored                            --  1 0x00 byte
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:153]: This `0x00` byte was [hacked in](https://github.com/fortanix/sdkms-plugin-library/pull/8/files#diff-99dc3f06dd4742e3f7d914953fbdc77546009acc6f3b1896beea3495bd7812fcL249). It is simpler to just ignore it and comply with it on serialization.

File: hd-wallet/plugin.lua [hunk 5]
```diff
@@ -1,440 +1,418 @@
...
-        local comp_key_2 = Blob.from_hex(parent_key.key)
-        local pt = secP256K1:point_from_binary(comp_key_2)
-        local x2 = pt:x()
-        local y2 = pt:y()
-        local p1 = secP256K1:point_from_components(x1, y1)
-        local p2 = secP256K1:point_from_components(x2, y2)
-        local p3 = p1 + p2
-        
-        child_key.key = compress_public_key(p3:x():to_bytes_be():hex(), p3:y():to_bytes_be():hex())
+local function parse_path(path)
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:206]: This function returns an array with the indices for sequential derivation
`m/0/0H/42/42H --> [nil, 0, 2 ** 31, 42, 42 + 2 ** 31]`
(it was part of the original plugin).

File: hd-wallet/plugin.lua [hunk 6]
```diff
@@ -1,440 +1,418 @@
...
-  if input.import == true then
-    return run_import(input)
-  end
-
-  if input.debug == true then
-    if input.coin == "eth" then return { result = run_eth(input); debug = DEBUG }
-    elseif input.coin == "xrp" then return { result = run_xrp(input); debug = DEBUG }
-    elseif input.coin == "utxo" then return { result = run_utxo(input); debug = DEBUG }
-    else return { error = "unsupported coin" } end
+  local required_fields = { "master_key_id", "msg_hash" }
```
Comment (zugzwang) [inline hd-wallet/plugin.lua:385]: - [X] Add coin

File: hd-wallet/test_vectors.json [hunk 7]
```diff
@@ -0,0 +1,113 @@
+[
```
Comment (zugzwang) [inline hd-wallet/test_vectors.json:1]: Test vectors from the spec.

**Comment by Nerotool** [comment]
SEND A DM TO NEROTOOL_ ON INSTAGRAM FOR HACKING

**Comment by zugzwang** [comment]
Closed in favour of https://github.com/fortanix/sdkms-plugin-library/commit/c194acaecc1bfdcbcc09cd7d6e646d8b8fe55bba

**Pull request closed without merging**